File: 2_Programação_sistemas_operacionais/2.4_Pipes_Sinais_Alarmes/Ex3.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <string.h>

int main()
{
	int pid;
	int fd[2];
	pipe(fd);
	pid = fork();
	if(pid == 0) // Codigo do filho
	{
		char buffer_filho[30], msg_filho[30] = "FILHO DIZ: HELLO PIPE";
		printf("Filho vai ler o pipe\n");
		if(read(fd[0], buffer_filho, 30) < 0) 
			printf("Erro na leitura do pipe\n");
		else
			printf("Valor lido pelo filho = %s\n", buffer_filho);
		printf("Filho vai escrever no pipe\n");
		if (write(fd[1], msg_filho, 30) < 0)
			printf("Erro na escrita do pipe\n");
		printf("Filho terminou de escrever no pipe\n");
	}
	else // Codigo do pai
	{
		char buffer_pai[30], msg_pai[30] = "PAI DIZ: HELLO PIPE";
		printf("Pai vai escrever no pipe\n");
		if (write(fd[1], msg_pai, 30) < 0)
			printf("Erro na escrita do pipe\n");
		printf("Pai terminou de escrever no pipe\n");
		printf("Pai vai hibernar por 1 segundo, para dar tempo do filho ler o pipe\n");
		sleep(1);
		printf("Pai vai ler o pipe\n");
		if(read(fd[0], buffer_pai, 30) < 0) 
			printf("Erro na leitura do pipe\n");
		else
			printf("Valor lido pelo pai = %s\n", buffer_pai);
	}
	return 0;
}

File: 2_Programação_sistemas_operacionais/2.4_Pipes_Sinais_Alarmes/Python/Ex3.py
import os, time

def preenche_str(s, n_bytes):
	s += '\0'*(n_bytes-len(s))
	return str.encode(s)

fd = os.pipe()
pid_filho = os.fork()
if pid_filho == 0:
	print("Filho vai ler o pipe")
	buffer_filho = os.read(fd[0], 30).decode("utf-8")
	print("Valor lido pelo filho = " + buffer_filho)
	print("Filho vai escrever no pipe")
	msg_filho = preenche_str("FILHO DIZ: HELLO PIPE", 30)
	os.write(fd[1], msg_filho)
	print("Filho terminou de escrever no pipe")
else:
	msg_pai = preenche_str("PAI DIZ: HELLO PIPE", 30)
	print("Pai vai escrever no pipe")
	os.write(fd[1], msg_pai)
	print("Pai terminou de escrever no pipe")
	print("Pai vai hibernar por 1 segundo, para dar tempo do filho ler o pipe")
	time.sleep(1)
	print("Pai vai ler o pipe")
	buffer_pai = os.read(fd[0], 30).decode("utf-8")
	print("Valor lido pelo pai = " + buffer_pai)